Add support for json output of results

Redmar-van-den-Berg opened the pull request:
## Pull Request Details
This pull request adds a new output file with command line option, `-j`, which contains the results for each marker in json format. This makes it easier to perform downstream processing on the TSSV results.

## Breaking Changes
None

## Issues Fixed
None

## Other Relevant Information
Below is a comparison between the data from the regular, and the new json report.

### The regular report from tssv
```
known alleles for marker chrM:152:
allele  total   forward reverse

new alleles for marker chrM:152 (mean length 1.0):
allele  total   forward reverse
C       27      8       19

known alleles for marker chrM:263:
allele  total   forward reverse

new alleles for marker chrM:263 (mean length 1.0):
allele  total   forward reverse
G       21      3       18

known alleles for marker chrM:302:
allele  total   forward reverse

new alleles for marker chrM:302 (mean length 3.111111111111111):
allele  total   forward reverse
ACC     16      0       16
ACCC    2       0       2
```
### The json report
```json
{
 "chrM:152": {
  "known": [],
  "new": [
   {
    "allele": "C",
    "total": 27,
    "forward": 8,
    "reverse": 19
   }
  ]
 },
 "chrM:263": {
  "known": [],
  "new": [
   {
    "allele": "G",
    "total": 21,
    "forward": 3,
    "reverse": 18
   }
  ]
 },
 "chrM:302": {
  "known": [],
  "new": [
   {
    "allele": "ACC",
    "total": 16,
    "forward": 0,
    "reverse": 16
   },
   {
    "allele": "ACCC",
    "total": 2,
    "forward": 0,
    "reverse": 2
   }
  ]
 }
```

[1]: https://github.com/jfjlaros/tssv/pulls
[2]: https://github.com/jfjlaros/tssv/blob/master/docs/CONTRIBUTING.md#code-style

Redmar-van-den-Berg committed:
Add support for json output of results

jfjlaros commented on tssv/tssv.py:
```diff
@@ -1,5 +1,6 @@
 from collections import defaultdict
 from functools import reduce
+import json
```
Could you use the `from .. import ..` syntax?

jfjlaros commented on tssv/tssv.py:
```diff
@@ -397,3 +417,4 @@ def tssv(
         write_files(tables, files)
 
     make_report(tables, report_handle)
+    make_json(tables, json_handle)
```
Would it not be better to make the output format a parameter (e.g., text, json or none)?

jfjlaros commented on tssv/cli.py:
```diff
@@ -31,6 +31,9 @@ def main():
         help='name of the report file')
+    parser.add_argument(
+        '-j', dest='json_handle', type=FileType('w'),
+        help='name of the json output file')
```
This parameter is optional, but it does not have a default.

jfjlaros reviewed (changes requested):
Could you have a look at these comments?

Redmar-van-den-Berg committed:
Make json report complete

Make sure that all the data that is present in the text report is also
present in the json report.

Redmar-van-den-Berg committed:
Add format option for report

Redmar-van-den-Berg committed:
Rename some functions

Redmar-van-den-Berg committed:
Group json report by marker instead of section

Instead of mimicking the format of the text report in json, the
information from both the 'allele' and 'library' sections are added
under the name of the individual markers. This should make working with
data easier, since all the data for each marker is in the same place.

Redmar-van-den-Berg committed:
Import from json

Redmar-van-den-Berg commented:
I added an option to specify the format of the report ('text' or 'json'). If no option is specified, the text version of the report is printed. 

I also changed the format of the json report to organise the results by marker, instead of by section. See the [attached files](https://github.com/jfjlaros/tssv/files/6767679/report.zip) for details.

Redmar-van-den-Berg committed:
Clean the allele field to only contain the allele

Redmar-van-den-Berg committed:
Dont use sub-function

Redmar-van-den-Berg committed:
Remove leftover comment

Redmar-van-den-Berg committed:
Expand the allele as specified by the occurrence

Redmar-van-den-Berg committed:
Add support for expanding more complex alleles

Redmar-van-den-Berg committed:
Leave known alleles as-is, instead of expanding

The known allele can be a complex, which makes expanding it difficult.
Also, by not expanding the known alleles, the difference between the
text and json reports are minimised.

Redmar-van-den-Berg commented:
The 'known' allele is left as-is in the json report, so that it matches the text report.

jfjlaros commented on tssv/cli.py:
```diff
@@ -31,6 +31,9 @@ def main():
         '-r', dest='report_handle', type=FileType('w'), default=stdout,
         help='name of the report file')
+    parser.add_argument(
+        '-f', dest='report_format', type=str, choices = ['text', 'json'],
```
One last remark, perhaps it would be better to change the type of this option to `bool`, e.g., make the option '-j' for a JSON report and omit for a text report.

Redmar-van-den-Berg committed:
Use boolean switch to enable json format report
